File: step_10_todos/harness/skills.py
"""Stage 10 - skills, unchanged since stage 9.
"""

from pathlib import Path

import yaml

SKILL_DIRS = [
    Path.home() / ".agents" / "skills",  # your skills
    Path.cwd() / ".agents" / "skills",   # this project's skills
]


def find_skills():
    """Glob SKILL.md under every skill dir; name -> {description, path}."""
    skills = {}
    for directory in SKILL_DIRS:
        for path in sorted(directory.glob("*/SKILL.md")):
            text = path.read_text(encoding="utf-8")
            if not text.startswith("---"):
                continue
            _, frontmatter, _ = text.split("---", 2)
            meta = yaml.safe_load(frontmatter) or {}
            if "name" not in meta:
                continue
            description = " ".join(str(meta.get("description", "")).split())
            skills[meta["name"]] = {"description": description, "path": path}
    return skills


SKILLS = find_skills()


def skills_prompt():
    """One line per skill: the index that goes into the system prompt."""
    return "\n".join(f"- {name}: {s['description']}" for name, s in SKILLS.items())


def read_skill(name: str) -> str:
    """Open a skill and return its full instructions."""
    if name not in SKILLS:
        return f"No skill named '{name}'."
    return SKILLS[name]["path"].read_text(encoding="utf-8")


if __name__ == "__main__":
    print(skills_prompt())

File: harness/06_production/step_45_typescript_core/harness-ts/skills.ts
/** Step 45 - skills: the same SKILL.md files the Python harness reads.
 *
 * The front matter is read with a small key: value reader, not a YAML
 * library. Every shipped skill keeps name and description on one line each,
 * and that is all the index needs.
 */

import { existsSync, readdirSync, readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

export const SKILL_DIRS = [
  join(homedir(), ".agents", "skills"), // your skills
  join(process.cwd(), ".agents", "skills"), // this project's skills
];

export type Skill = { description: string; path: string };

/** key: value lines, with indented continuation lines folded into the value above. */
export function frontmatter(text: string): Record<string, string> {
  const meta: Record<string, string> = {};
  let last: string | null = null;
  for (const line of text.split(/\r?\n/)) {
    const match = /^([A-Za-z_][\w-]*):\s*(.*)$/.exec(line);
    if (match) {
      last = match[1];
      meta[last] = match[2].replace(/^["'](.*)["']$/, "$1").replace(/^[>|]-?$/, "");
    } else if (last !== null && /^\s+\S/.test(line)) {
      meta[last] = `${meta[last]} ${line.trim()}`.trim();
    }
  }
  return meta;
}

/** Read SKILL.md under every skill dir; name -> {description, path}. */
export function findSkills(): Record<string, Skill> {
  const skills: Record<string, Skill> = {};
  for (const directory of SKILL_DIRS) {
    if (!existsSync(directory)) continue;
    for (const name of readdirSync(directory).sort()) {
      const path = join(directory, name, "SKILL.md");
      if (!existsSync(path)) continue;
      const text = readFileSync(path, "utf-8");
      if (!text.startsWith("---")) continue;
      const meta = frontmatter(text.split("---")[1] ?? "");
      if (!meta.name) continue;
      const description = (meta.description ?? "").split(/\s+/).join(" ").trim();
      skills[meta.name] = { description, path };
    }
  }
  return skills;
}

export const SKILLS = findSkills();

/** One line per skill: the index that goes into the system prompt. */
export function skillsPrompt(): string {
  return Object.entries(SKILLS)
    .map(([name, s]) => `- ${name}: ${s.description}`)
    .join("\n");
}

/** Open a skill and return its full instructions. */
export function readSkill({ name }: { name: string }): string {
  if (!(name in SKILLS)) return `No skill named '${name}'.`;
  return readFileSync(SKILLS[name].path, "utf-8");
}
